nativelib/src/main/scala-3/scala/scalanative/safe/SafeZone.scala

```diff
@@ -0,0 +1,59 @@
+package scala.scalanative.safe
+
+import language.experimental.captureChecking
+import scalanative.unsigned._
+import scala.annotation.implicitNotFound
+import scala.scalanative.runtime.{RawPtr, CMemoryPool, CMemoryPoolZone}
+
+@implicitNotFound("Given method requires an implicit zone.")
+trait SafeZone {
+
+  /** Frees allocations. This zone allocator is not reusable once closed. */
+  def close(): Unit
+
+  /** Return this zone allocator is open or not. */
+  def isOpen: Boolean
+
+  /** Return this zone allocator is closed or not. */
+  def isClosed: Boolean = !isOpen
+
+  /** Return the handle of this zone allocator. */
+  def handle: RawPtr
+
+  /** Require this zone allocator is open. */
+  def checkOpen: Unit = {
+    if (!isOpen)
+      throw new IllegalStateException(s"Zone ${this} is already closed.")
+  }
+}
+
+final class MemorySafeZone (private[this] val zoneHandle: RawPtr) extends SafeZone {
```

Comment:
If we go with the previous comment we can simplify it a bit
```suggestion
final class MemorySafeZone (private[scalanative] val handle: RawPtr) extends SafeZone {
```
and remove the public accessor